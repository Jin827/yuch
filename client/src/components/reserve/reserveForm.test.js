import React from 'react';
import { shallow } from 'enzyme';
import { unwrap } from '@material-ui/core/test-utils';
import ReserveForm from './reserveForm';
import * as mockData from '../../__tests__/__mocks__/mockData';

describe('<ReserveForm />', () => {
  const mockClose = jest.fn();
  const mockChange = jest.fn();
  const mockSubmit = jest.fn();

  const ComponentNaked = unwrap(ReserveForm);

  const props = {
    inThreeDays: mockData.inThreeDays,
    handleClose: mockClose,
    handleChange: mockChange,
    handleSubmit: mockSubmit,
    submitBtnClicked: false,
    classes: {},
  };

  describe('with initial reserve info', () => {
    const wrapper = shallow(
      <ComponentNaked {...props} reserveInfo={mockData.reserveInfoInit} />,
    );

    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('creates inputs', () => {
      expect(wrapper.find('.input-name').exists()).toBe(true);
      expect(wrapper.find('.input-contact').exists()).toBe(true);
      expect(wrapper.find('.input-number').exists()).toBe(true);
      expect(wrapper.find('.input-place').exists()).toBe(true);
      expect(wrapper.find('.input-date').exists()).toBe(true);
      expect(wrapper.find('.input-time').exists()).toBe(true);
    });

    describe('when typeing into input, calls a callback function', () => {
      it('name', () => {
        const ev = { target: { id: 'name', value: 'Yuchung' } };
        wrapper.find('.input-name').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });

      it('contact', () => {
        const ev = { target: { id: 'contact', value: '123123123' } };
        wrapper.find('.input-contact').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });

      it('number', () => {
        const ev = { target: { id: 'number', value: '100' } };
        wrapper.find('.input-number').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });

      it('place', () => {
        const ev = { target: { id: 'place', value: 'Gyeongju' } };
        wrapper.find('.input-place').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });

      it('date', () => {
        const ev = { target: { id: 'date', value: '2019-12-25' } };
        wrapper.find('.input-date').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });

      it('time', () => {
        const ev = { target: { id: 'time', value: '2:30 PM' } };
        wrapper.find('.input-time').simulate('change', ev);
        expect(mockChange).toHaveBeenCalledWith(ev);
      });
    });

    // describe('clicking a button', () => {
    //   describe('submit button', () => {
    //     const mockedEvent = { target: {} };
    //     beforeEach(() => {
    //       wrapper.find('.btn-submit').simulate('click', mockedEvent);
    //     });

    //     it('handle submit function', () => {
    //       expect(mockSubmit).toHaveBeenCalledWith(mockedEvent);
    //     });
    //   });

    //   it('close button', () => {
    //     wrapper.find('.btn-close').simulate('click');
    //     expect(mockClose).toHaveBeenCalledWith();
    //   });
    // });
  });

  describe('throwing error for an empty input value on submit', () => {
    // const mockedEvent = { target: {} };

    describe('should show error when input field is empty', () => {
      const setup = () => {
        const wrapper = shallow(
          <ComponentNaked
            {...props}
            reserveInfo={{
              name: '',
              contact: '(0  )    -    ',
              number: '',
              place: '',
              date: '',
              time: '',
              createdAt: '',
            }}
          />,
        );
        return { wrapper };
      };

      const { wrapper } = setup();

      beforeEach(() => {
        // wrapper.find('.btn-submit').simulate('click', mockedEvent);
        wrapper.setProps({ submitBtnClicked: true });
      });

      it('when name field is empty', () => {
        expect(
          wrapper
            .find('.input-name')
            .first()
            .props().error,
        ).toEqual(true);
      });

      it('when number field is empty', () => {
        expect(
          wrapper
            .find('.input-number')
            .first()
            .props().error,
        ).toEqual(true);
      });

      it('when place field is empty', () => {
        expect(
          wrapper
            .find('.input-place')
            .first()
            .props().error,
        ).toEqual(true);
      });

      it('when time field is empty', () => {
        expect(
          wrapper
            .find('.input-time')
            .first()
            .props().error,
        ).toEqual(true);
      });

      it('when date field is empty', () => {
        expect(
          wrapper
            .find('.input-date')
            .first()
            .props().error,
        ).toEqual(true);
      });
    });
  });
});
