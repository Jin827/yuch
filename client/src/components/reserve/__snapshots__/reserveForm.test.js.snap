// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReserveForm /> with initial reserve info renders correctly 1`] = `
<Fragment>
  <div
    className="mh2"
  >
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      className="input-name"
      error={false}
      fullWidth={true}
      id="name"
      label="이름"
      margin="normal"
      onChange={[Function]}
      required={true}
      select={false}
      value=""
      variant="standard"
    />
    <WithStyles(FormControl)
      fullWidth={true}
      margin="normal"
    >
      <WithStyles(InputLabel)
        htmlFor="formatted-text-mask-input"
      >
        연락처 *
      </WithStyles(InputLabel)>
      <WithStyles(Input)
        className="input-contact"
        error={false}
        id="contact"
        inputComponent={[Function]}
        onChange={[Function]}
        placeholder="(054)-745-0999"
        required={true}
        value="(0  )    -    "
      />
    </WithStyles(FormControl)>
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      className="input-number"
      error={false}
      fullWidth={true}
      id="number"
      label="인원수"
      margin="normal"
      onChange={[Function]}
      required={true}
      select={false}
      type="number"
      value=""
      variant="standard"
    />
    <TextField
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      className="input-place"
      error={false}
      fullWidth={true}
      helperText="경주시 내 전지역 배달 가능합니다."
      id="place"
      label="장소"
      margin="normal"
      multiline={true}
      onChange={[Function]}
      required={true}
      select={false}
      value=""
      variant="standard"
    />
    <TextField
      className="input-date"
      defaultValue="1970-01-04"
      error={false}
      fullWidth={true}
      helperText="최소 3일 전일 경우에만 예약 진행 가능합니다."
      id="date"
      label="날짜"
      margin="normal"
      onChange={[Function]}
      required={true}
      select={false}
      type="date"
      variant="standard"
    />
    <TextField
      className="input-time"
      defaultValue="12:30"
      error={false}
      fullWidth={true}
      helperText="( 예약 가능 시간 )  11:00 - 20:00"
      id="time"
      inputProps={
        Object {
          "step": 900,
        }
      }
      label="시간"
      margin="normal"
      onChange={[Function]}
      required={true}
      select={false}
      type="time"
      variant="standard"
    />
  </div>
  <div
    className="mt3"
  >
    <WithStyles(Button)
      className="btn-submit undefined"
      color="secondary"
      onClick={[Function]}
      variant="contained"
    >
      예약완료
    </WithStyles(Button)>
    <WithStyles(Button)
      className="btn-close undefined"
      color="secondary"
      onClick={[MockFunction]}
      variant="outlined"
    >
      닫기
    </WithStyles(Button)>
  </div>
</Fragment>
`;
